File: DSA/RATNA_KHANNA_590013141/DAY_75_QUESTION_2.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* left;
    struct Node* right;
};

struct Node* newNode(int data) {
    struct Node* node = (struct Node*)malloc(sizeof(struct Node));
    node->data = data;
    node->left = node->right = NULL;
    return node;
}

struct Queue {
    struct Node* arr[100];
    int front, rear;
};

void initQueue(struct Queue* q) {
    q->front = q->rear = -1;
}

int isEmpty(struct Queue* q) {
    return q->front == -1;
}

void enqueue(struct Queue* q, struct Node* node) {
    if (q->rear == 99) return;
    if (q->front == -1) q->front = 0;
    q->arr[++q->rear] = node;
}

struct Node* dequeue(struct Queue* q) {
    if (isEmpty(q)) return NULL;
    struct Node* node = q->arr[q->front];
    if (q->front == q->rear) q->front = q->rear = -1;
    else q->front++;
    return node;
}

void rightView(struct Node* root) {
    if (root == NULL) return;
    struct Queue q;
    initQueue(&q);
    enqueue(&q, root);
    while (!isEmpty(&q)) {
        int n = q.rear - q.front + 1;
        for (int i = 0; i < n; i++) {
            struct Node* temp = dequeue(&q);
            if (i == n - 1) printf("%d ", temp->data);
            if (temp->left) enqueue(&q, temp->left);
            if (temp->right) enqueue(&q, temp->right);
        }
    }
}

struct Node* buildTree() {
    int x;
    scanf("%d", &x);
    if (x == -1) return NULL;
    struct Node* node = newNode(x);
    printf("Enter left child of %d (-1 for no child): ", x);
    node->left = buildTree();
    printf("Enter right child of %d (-1 for no child): ", x);
    node->right = buildTree();
    return node;
}

int main() {
    printf("Enter root value (-1 for no node): ");
    struct Node* root = buildTree();
    printf("Right View of Binary Tree: ");
    rightView(root);
    return 0;
}

File: DSA/LakshaKundra_590013237/Day_75_Question_2.py
from collections import deque

class Node:
    def __init__(self, key):
        self.data = key
        self.left = None
        self.right = None

def right_view(root):
    if not root:
        return
    q = deque([root])
    while q:
        n = len(q)
        for i in range(n):
            node = q.popleft()
            if i == n - 1:
                print(node.data, end=" ")
            if node.left:
                q.append(node.left)
            if node.right:
                q.append(node.right)

# Example 1
root = Node(1)
root.left = Node(2)
root.right = Node(3)
root.left.left = Node(4)

right_view(root)  # Output: 1 3 4

print()

# Example 2
root = Node(10)
root.left = Node(20)
root.right = Node(30)
root.left.left = Node(40)
root.left.right = Node(50)
root.right.left = Node(60)
root.right.right = Node(70)

right_view(root)  # Output: 10 30 70
